### PR TITLE
soffice: Announce actually deleted text on Ctrl+backspace

### LINKED ISSUE #15436: The ctrl-backspace command doesn't read the deleted word in LibreOffice Writer
### Steps to reproduce:

Open LibreOffice Writer. I have tested with version 7.6.0.6 64 bits, but I don't think it especially matters:

Write a few words, for example: NVDA is a free screen reader.

Then press control-backspace a couple of times.

### Actual behavior:

````
Untitled 1 — LibreOffice Writer
Untitled 1 - LibreOffice Document  document  editable
blank
speak command keys on
NVDA
is
a
free
screen
reader
ctrl+backspace
dot
ctrl+backspace
reader
ctrl+backspace
space
ctrl+backspace
space
ctrl+backspace
space
ctrl+backspace
space
ctrl+backspace
space
ctrl+backspace

````

The dot and the word reader were correctly announced, but not the others.

### Expected behavior:

````
ctrl+backspace
dot
ctrl+backspace
reader
ctrl+backspace
screen
ctrl+backspace
free
ctrl+backspace
a
ctrl+backspace
is
ctrl+backspace
sNVDA
ctrl+backspace

````

Each press of ctrl-backspace should announce the word that is being deleted.

#### NVDA installed/portable/running from source:
Installed.

#### NVDA version:
2023.2

#### Windows version:
Windows 10 22H2 (AMD64) build 19045.3324

#### Name and version of other software in use when reproducing the issue:

````
Version: 7.6.0.3 (X86_64) / LibreOffice Community
Build ID: 69edd8b8ebc41d00b4de3915dc82f8f0fc3b6265
CPU threads: 8; OS: Windows 10.0 Build 19045; UI render: Skia/Raster; VCL: win
Locale: en-GB (es_ES); UI: en-GB
Calc: CL threaded
````

#### Does the issue still occur after restarting your computer?
Yes.

#### Have you tried any other versions of NVDA? If so, please report their behaviors.
I have not.

#### If NVDA add-ons are disabled, is your problem still occurring?
Yes.

#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
Yes.

### PR DESCRIPTION
**Pull request by michaelweghorn**
### Link to issue number:

Fixes #15436

 ### Summary of the issue:

The base class implementation
`EditableText#_backspaceScriptHelper` for handling the Ctrl+Backspace keyboard shortcut retrieves the text to announce as deleted by expanding the text info to the corresponding text unit and then retrieving the text. This gives an incorrect result in at least Libreoffice Writer, where that would return a string just containing the space character when removing a word followed by a space using the Ctrl+backspace keyboard shortcut.
(Writer's `IAccessibleText::textAtOffset` implementation that gets used in `IA2TextTextInfo#_getWordOffsets` handles the space as a separate word, but Ctrl+backspace removes both, the actual word and following whitespace.)

 ### Description of user facing changes

When removing a word followed by whitespace in Libreoffice Writer, the actually removed text is announced.

 ### Description of development approach

Instead of separately retrieving the text for a word, override the default implementation in the LibreOffice app module and announce the actually removed text when handling Ctrl+backspace.
Use the caret position before and after the text has been removed to identify what the removed text is.

 ### Testing strategy:

Test the scenario as described in issue #15436:

1) start LibreOffice Writer
2) type "NVDA is a free screen reader."
3) press Ctrl+backspace 7 times
4) verify that the actually removed text is annonced by NVDA.

 ### Known issues with pull request:

None

 ### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

### CODE REVIEW

**Comment by OzancanKaratas** [comment]
Please do not touch the changes file. This file will be changed by NV Access.

**Comment by michaelweghorn** [comment]
> Please do not touch the changes file. This file will be changed by NV Access.

There was this recent email on the nvda-devel mailing list about a change on how this is handled:
https://groups.io/g/nvda-devel/message/46876

My understanding is that the changes file should be modified directly now (which this change does), or do I misunderstand anything?

**Comment by OzancanKaratas** [comment]
No. Do not revert. I misled you because I was not aware of the final decision. I am sorry.

File: source/appModules/soffice.py [hunk 1]
```diff
@@ -318,6 +318,43 @@ def _get_locationText(self):
 class SymphonyDocument(CompoundDocument):
 	TextInfo = SymphonyDocumentTextInfo
 
+	# override base class implementation because that one assumes
+	# that the text retrieved from the text info for the text unit
+	# is the same as the text that actually gets removed, which at
+	# least isn't true for Writer paragraphs when removing a word
+	# followed by whitespace using Ctrl+Backspace
+	def _backspaceScriptHelper(self, unit, gesture):
```
Comment (seanbudd) [inline source/appModules/soffice.py:326]: could this function please get type hints added?

```suggestion
	def _backspaceScriptHelper(self, unit: str, gesture: inputCore.InputGesture):
```

File: source/appModules/soffice.py [hunk 2]
```diff
@@ -318,6 +318,43 @@ def _get_locationText(self):
...
+	# override base class implementation because that one assumes
+	# that the text retrieved from the text info for the text unit
+	# is the same as the text that actually gets removed, which at
+	# least isn't true for Writer paragraphs when removing a word
+	# followed by whitespace using Ctrl+Backspace
+	def _backspaceScriptHelper(self, unit, gesture):
+		try:
+			oldInfo = self.makeTextInfo(textInfos.POSITION_CARET)
+			if not isinstance(oldInfo, SymphonyDocumentTextInfo):
+				super(SymphonyDocument, self)._backspaceScriptHelper(unit, gesture)
```
Comment (seanbudd) [inline source/appModules/soffice.py:330]: Py3 syntax change

```suggestion
				super()._backspaceScriptHelper(unit, gesture)
```

File: user_docs/en/changes.t2t [hunk 3]
```diff
@@ -16,6 +16,7 @@ What's New in NVDA
 - The following commands now support two and three presses to spell the reported information and spell with character descriptions: report selection, report clipboard text and report focused object. (#15449)
 - The option "Report role when mouse enters object" in NVDA's mouse settings category has been renamed to "Report object when mouse enters it".
 This option now announces additional relevant information about an object when the mouse enters it, such as states (checked/pressed) or cell coordinates in a table. (#15420)
+- In LibreOffice, words deleted using the Ctrl+Backspace keyboard shortcut are now also properly announced when the deleted word is followed by whitespace (like spaces and tabs). (#15436)
```
Comment (seanbudd) [inline user_docs/en/changes.t2t:19]:
```suggestion
- In LibreOffice, words deleted using the ``control+backspace`` keyboard shortcut are now also properly announced when the deleted word is followed by whitespace (like spaces and tabs). (#15436)
```

**Review by seanbudd — COMMENTED** [review]
Thanks @michaelweghorn

File: source/appModules/soffice.py (hunk 1 above)
Comment (michaelweghorn) [reply source/appModules/soffice.py]: Done in the new version.

File: source/appModules/soffice.py (hunk 2 above)
Comment (michaelweghorn) [reply source/appModules/soffice.py]: Done in the new version.

File: user_docs/en/changes.t2t (hunk 3 above)
Comment (michaelweghorn) [reply user_docs/en/changes.t2t]: Done in the new version.

File: source/appModules/soffice.py [hunk 4]
```diff
@@ -318,6 +319,43 @@ def _get_locationText(self):
...
 
+	# override base class implementation because that one assumes
+	# that the text retrieved from the text info for the text unit
+	# is the same as the text that actually gets removed, which at
+	# least isn't true for Writer paragraphs when removing a word
+	# followed by whitespace using Ctrl+Backspace
+	def _backspaceScriptHelper(self, unit: str, gesture: inputCore.InputGesture):
+		try:
+			oldInfo = self.makeTextInfo(textInfos.POSITION_CARET)
+			if not isinstance(oldInfo, SymphonyDocumentTextInfo):
```
Comment (lukaszgo1) [inline source/appModules/soffice.py:330]: It seems text info for this class is explicitly set to `SymphonyDocumentTextInfo`, so the `isinstance` call seems unnecessary. Have you encountered a case in which text info was of a different type here?

Comment (michaelweghorn) [reply source/appModules/soffice.py]: No, I haven't. Looking at it again, that check is indeed unnecessary and I've dropped it now. Thanks!

File: user_docs/en/changes.t2t [hunk 5]
```diff
@@ -16,6 +16,7 @@ What's New in NVDA
 - The following commands now support two and three presses to spell the reported information and spell with character descriptions: report selection, report clipboard text and report focused object. (#15449)
 - The option "Report role when mouse enters object" in NVDA's mouse settings category has been renamed to "Report object when mouse enters it".
 This option now announces additional relevant information about an object when the mouse enters it, such as states (checked/pressed) or cell coordinates in a table. (#15420)
+- In LibreOffice, words deleted using the ``control+backspace`` keyboard shortcut are now also properly announced when the deleted word is followed by whitespace (like spaces and tabs). (#15436)
```
Comment (seanbudd) [inline user_docs/en/changes.t2t:19]: could this be moved to bug fixes?

Comment (michaelweghorn) [reply user_docs/en/changes.t2t]: Yes, done in the newly pushed version.

**Comment by michaelweghorn** [comment]
I did a rebase onto current git master to fix the merge conflict in the changelog.

**Review by seanbudd — APPROVED** [review]
Thanks @michaelweghorn

**Pull request merged**